### ISSUE TITLE
The Starlette lifespan async context manager is not executed

### ISSUE THREAD

**Issue by ekaterina-nikonova** [zeit/now#11677]
When a Starlette application is run using Vercel, the body of the `lifespan` context manager is never executed. Here's a minimal `main.py` example illustrating the issue:

```python
from collections.abc import AsyncIterator
from contextlib import asynccontextmanager
from typing import TypedDict

from starlette.applications import Starlette
from starlette.requests import Request
from starlette.responses import PlainTextResponse
from starlette.routing import Route


async def root(request: Request):
    print("In the root")
    return PlainTextResponse("This is a test for Starlette lifespan with Vercel.")


class State(TypedDict):
    lifespan_key: str


@asynccontextmanager
async def lifespan(app: Starlette) -> AsyncIterator[State]:
    print("In the lifespan")
    yield {"lifespan_key": "lifespan_value"}


routes = [
    Route('/', endpoint=root),
]


app = Starlette(
    lifespan=lifespan,
    routes=routes,
)
```

The `print` statement in the `root` handler makes sure the output is not swallowed. When we run the app with Uvicorn, the output is as expected (first, the lifespan is executed, then -- after navigating to the `/` route -- the `root` handler is executed):

```bash
$ uvicorn main:app
INFO:     Started server process [30956]
INFO:     Waiting for application startup.
In the lifespan
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
In the root
INFO:     127.0.0.1:52827 - "GET / HTTP/1.1" 200 OK
```

When we run the same app with Vercel, we never get the output <em>In the lifespan</em> from the context manager, though the output from the `root` handler -- <em>In the root</em> -- is present:

```bash
$ vercel dev
(node:30994) [DEP0040] DeprecationWarning: The `punycode` module is deprecated. Please use a userland alternative instead.
(Use `node --trace-deprecation ...` to show where the warning was created)
(node:30994) [DEP0060] DeprecationWarning: The `util._extend` API is deprecated. Please use Object.assign() instead.
Vercel CLI 34.2.4
> Ready! Available at http://localhost:3000
> Building @vercel/python:main.py
(node:30998) [DEP0040] DeprecationWarning: The `punycode` module is deprecated. Please use a userland alternative instead.
(Use `node --trace-deprecation ...` to show where the warning was created)
Installing required dependencies...
> Built @vercel/python:main.py [404ms]
using Asynchronous Server Gateway Interface (ASGI)
In the root
```

Following the track all the way into Starlette's `routing.py`, I see that `Router.app()` is never called with `scope["type"] == "lifespan"`, as when using Uvicorn.

My `vercel.json` file:

```json
{
  "version": 2,
  "builds": [
    {
      "src": "main.py",
      "use": "@vercel/python"
    }
  ],
  "routes": [
    {
      "src": "(.*)",
      "dest": "main.py"
    }
  ]
}
```

**Comment by andersonkxiass** [comment]
Hi @ekaterina-nikonova did you managed to figure out how to fix that ?

**Comment by rupinder2** [comment]
any updates on this?

**Comment by andersonkxiass** [comment]
After investigating this for a while, I believe I found the answer. This is not a specific "issue" with Vercel. I tried doing the same thing with other serverless providers and received the same result.

The reason the `lifespan` event doesn't work as expected in a serverless environment is due to how serverless platforms fundamentally operate.

In **traditional** servers, such as VMs, our application starts once, runs continuously, handles multiple requests, and keeps running until it is explicitly stopped.

In a **serverless** environment, application instances are created and destroyed frequently, with no guarantee of continuous runtime. Instances can be "frozen" or "thawed" at any time.

That being said, as an alternative to using `lifespan`, we might want to consider using `lru_cache` to manage objects.

**Comment by OR13** [comment]
This is sorta sad, because:

https://github.com/modelcontextprotocol/python-sdk

```python
# math.py
from mcp.server.fastmcp import FastMCP

mcp = FastMCP(name="MathServer", stateless_http=True)


@mcp.tool(description="A simple add tool")
def add_two(n: int) -> int:
    return n + 2
```

```python
# main.py
import contextlib
from fastapi import FastAPI
from mcp.echo import echo
from mcp.math import math


# Create a combined lifespan to manage both session managers
@contextlib.asynccontextmanager
async def lifespan(app: FastAPI):
    async with contextlib.AsyncExitStack() as stack:
        await stack.enter_async_context(echo.mcp.session_manager.run())
        await stack.enter_async_context(math.mcp.session_manager.run())
        yield


app = FastAPI(lifespan=lifespan)
app.mount("/echo", echo.mcp.streamable_http_app())
app.mount("/math", math.mcp.streamable_http_app())
```

**Comment by sjsadowski** [comment]
Dropping in here after banging my head against the wall for a few days around this and wondering why I was running into issues with lack of connections to neon and redis in dev/preview environments.

I feel like this is only important in cases where the lifecycle is important, and honestly in those cases (especially with database setup/teardown being part of the lifecycle for a reason) there should be a way to avoid suspension and opt for a shutdown and a cold boot when necessary.

If nothing else it should be documented that the contextmanager is ignored, but in reality dealing with properly handling database connections on every request is a lot of overhead that could increase possibilities of errors due to connection throttling or just general RTT because the teardown would have to be handled before the response could be sent to the client. 

**tldr**: if we can avoid suspend/resume and optionally handle a shutdown gracefully for the function being idle, then there's no reason to avoid the lifecycle context and it would reduce the overhead tax of things that should not be run every time a request comes in.

**Comment by sjsadowski** [comment]
To solve this problem, I added middleware that checks whether or not the lifespan is completed, and if not sets up and tears down the database connections every time, here's the difference between the two:

Without lifespan context:
```sh
(venv:local) [💻 ssadowski@goldfinger|(vercel-api)|sjsadowski.com-next]$ timeit curl -s -H 'Authorization: Bearer [token_redacted]' https://localhost:5443/api/healthcheck
589.812334ms

(venv:local) [💻 ssadowski@goldfinger|(vercel-api)|sjsadowski.com-next]$ timeit curl -s -H 'Authorization: Bearer [token_redacted]' https://localhost:5443/api/healthcheck 
961.980209ms

(venv:local) [💻 ssadowski@goldfinger|(vercel-api)|sjsadowski.com-next]$ timeit curl -s -H 'Authorization: Bearer [token_redacted]' https://localhost:5443/api/healthcheck
544.239084ms
```

With lifespan context:
```sh
(venv:local) [💻 ssadowski@goldfinger|(vercel-api)|sjsadowski.com-next]$ timeit curl -s -H 'Authorization: Bearer [token_redacted]' https://localhost:5443/api/healthcheck
72.672959ms

(venv:local) [💻 ssadowski@goldfinger|(vercel-api)|sjsadowski.com-next]$ timeit curl -s -H 'Authorization: Bearer [token_redacted]' https://localhost:5443/api/healthcheck
22.349125ms

(venv:local) [💻 ssadowski@goldfinger|(vercel-api)|sjsadowski.com-next]$ timeit curl -s -H 'Authorization: Bearer [token_redacted]' https://localhost:5443/api/healthcheck 
19.80075ms
```

Now obviously this doesn't fix a cold start and doesn't purport to, however the difference is clear: .5-1 second for the RTT without the context manager, .07 down to .02 seconds with it. Now for my own pet project, this is probably fine. And can likely be optimized to some extent, but it seems to me this is a problem worth solving.
